presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java

```diff
@@ -0,0 +1,148 @@
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `--...
+ *
+ * This will NOT merge the functions in WindowNodes that have identical WindowNode.Specifications,
+ * but have a node between them that is not a WindowNode.
+ * In the following example, the functions in the WindowNodes with specification `A' will not be
+ * merged into a single WindowNode.
+ *
+ * OutputNode
+ * `--...
+ *    `--WindowNode(Specification: A, Functions: [sum(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `-- ProjectNode(...)
+ *             `--WindowNode(Specification: A, Functions: [avg(something)])
+ *                `--...
+ */
+public class MergeIdenticalWindows
+        implements PlanOptimizer
+{
+    public PlanNode optimize(PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        return SimplePlanRewriter.rewriteWith(new Rewriter(), plan, new Context());
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<MergeIdenticalWindows.Context>
+    {
+        @Override
+        public PlanNode visitWindow(
+                WindowNode node,
+                RewriteContext<MergeIdenticalWindows.Context> context)
+        {
+            Context mergeContext = context.get();
+            boolean retainNode = mergeContext.collectFunctions(node);
+
+            // Don't merge window functions across nodes that aren't WindowNodes
+            if (node.getClass().isInstance(node.getSource())) {
+                node = (WindowNode) context.defaultRewrite(node, mergeContext);
+            }
+            else {
+                node = (WindowNode) context.defaultRewrite(node, new Context());
+            }
+
+            // WindowFunctions contained in this node will be merged in some node further up the tree.
+            if (!retainNode) {
+                return node.getSource();
+            }
+
+            Context.FunctionPair collectedFunctions = mergeContext.getCollectedFunctions(node);
+
+            return new WindowNode(
+                    node.getId(),
+                    node.getSource(),
+                    node.getSpecification(),
+                    collectedFunctions.windowFunctions.build(),
+                    collectedFunctions.signatures.build(),
+                    node.getHashSymbol(),
+                    node.getPrePartitionedInputs(),
+                    node.getPreSortedOrderPrefix());
+        }
+    }
+
+    private static class Context
+    {
+        private Map<WindowNode.Specification, FunctionPair> pairMap = new HashMap<>();
+
+        private boolean collectFunctions(WindowNode window)
+        {
+            WindowNode.Specification key = window.getSpecification();
+            FunctionPair pair = pairMap.get(key);
```

Comment:
That makes it impossible to know if the value is newly created, which we use to know if the WindowNode we're collecting functions from is the first one with a particular specification. I can take a look at changing this if it makes sense to rework the overall algorithm per your suggestion above.